src/compiler/binder.ts

```diff
@@ -1751,4 +1791,437 @@ namespace ts {
+                // These nodes are ES6 syntax.
+                transformFlags |= TransformFlags.AssertES6;
+                break;
+
+            case SyntaxKind.ObjectLiteralExpression:
+                excludeFlags = TransformFlags.ObjectLiteralExcludes;
+                if (subtreeFlags & TransformFlags.ContainsComputedPropertyName) {
+                    // If an ObjectLiteralExpression contains a ComputedPropertyName, then it
+                    // is an ES6 node.
+                    transformFlags |= TransformFlags.AssertES6;
+                }
+                break;
+
+            case SyntaxKind.CallExpression:
+                excludeFlags = TransformFlags.ArrayLiteralOrCallOrNewExcludes;
+                if (subtreeFlags & TransformFlags.ContainsSpreadElementExpression
+                    || isSuperCall(node)) {
+                    // If the this node contains a SpreadElementExpression, or is a super call, then it is an ES6
+                    // node.
+                    transformFlags |= TransformFlags.AssertES6;
+                }
+                break;
+
+            case SyntaxKind.ArrayLiteralExpression:
+            case SyntaxKind.NewExpression:
+                excludeFlags = TransformFlags.ArrayLiteralOrCallOrNewExcludes;
+                if (subtreeFlags & TransformFlags.ContainsSpreadElementExpression) {
+                    // If the this node contains a SpreadElementExpression, then it is an ES6
+                    // node.
+                    transformFlags |= TransformFlags.AssertES6;
+                }
+                break;
+
+            case SyntaxKind.Decorator:
+                // This node is TypeScript syntax, and marks its container as also being TypeScript syntax.
+                transformFlags |= TransformFlags.AssertTypeScript | TransformFlags.ContainsDecorators;
+                break;
+
+            case SyntaxKind.ModuleDeclaration:
+                // This node is TypeScript syntax, and excludes markers that should not escape the module scope.
+                excludeFlags = TransformFlags.ModuleExcludes;
+                transformFlags |= TransformFlags.AssertTypeScript;
+                break;
+
+            case SyntaxKind.ParenthesizedExpression:
+                // If the node is synthesized, it means the emitter put the parentheses there,
+                // not the user. If we didn't want them, the emitter would not have put them
+                // there.
+                if (!nodeIsSynthesized(node)) {
+                    if ((<ParenthesizedExpression>node).expression.kind === SyntaxKind.AsExpression
+                        || (<ParenthesizedExpression>node).expression.kind === SyntaxKind.TypeAssertionExpression) {
+                        transformFlags = TransformFlags.AssertTypeScript;
+                    }
+                }
+                break;
+
+            case SyntaxKind.ExpressionStatement:
+                // if (node.flags & NodeFlags.Generated) {

```

Comment:
I will have a later PR that uses this logic, but just didn't want it to get lost in the shuffle.